### ISSUE TITLE
Problem to resolve type by Scope

### ISSUE THREAD

**Issue by lillo42** [autofac/Autofac.AspNetCore.Multitenant#22]
In some case, Autofac is not creating a new instance, the type was register as scoped, after 2nd request for same tenant.

I'm using ASP NET Core 3.1 and for this sample the problem is to resolve ``` Test ```

A repo with sample to reproduce this bug: https://github.com/lillo42/autofac-multi-tenant

Program.cs:
```c#
public class Program
{
    public static void Main(string[] args)
    {
        CreateHostBuilder(args).Build().Run();
    }

    public static IHostBuilder CreateHostBuilder(string[] args) =>
        Host.CreateDefaultBuilder(args)
            .ConfigureAppConfiguration(config =>
            {
                config
                    .SetBasePath(AppContext.BaseDirectory)
                    .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
                    .AddJsonFile("appsettings.overrides.json", optional: true, reloadOnChange: true)
                    .AddEnvironmentVariables()
                    .AddCommandLine(args ?? new string[] { });
             })
             .UseServiceProviderFactory(new AutofacMultitenantServiceProviderFactory(Startup.ConfigureMultitenantContainer))
            //.UseServiceProviderFactory(new AutofacServiceProviderFactory())
            .ConfigureWebHostDefaults(webHostBuilder => webHostBuilder
                .UseContentRoot(AppContext.BaseDirectory)
                    .UseIISIntegration()
                    .UseKestrel()
                    .UseStartup<Startup>());

}
```

Startup.cs:
```c#
public class Program
{
    ...
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddMvc();
        services.AddControllers();

        services.AddScoped<ITest>(sp =>
        {
            return sp.GetRequiredService<Test>();
        });

        services.AddScoped<Test>();
        //services.AddScoped<ISampleDescription, SampleDescriptionGB>();
        services.AddAutofacMultitenantRequestServices();
    }
    ...

    public static MultitenantContainer ConfigureMultitenantContainer(IContainer container)
    {
        var strategy = new MyTenantIdentificationStrategy(container.Resolve<IHttpContextAccessor>());
        var mtc = new MultitenantContainer(strategy, container);

        //configure tenant dependencies here
        mtc.ConfigureTenant(1, cb => cb.RegisterType<SampleDescriptionGB>().As<ISampleDescription>().InstancePerLifetimeScope());
        mtc.ConfigureTenant(2, cb => cb.RegisterType<SampleDescriptionDE>().As<ISampleDescription>().InstancePerLifetimeScope());

        return mtc;
    }
}

public interface ITest { string GetId(); }

public class Test : ITest
{
    private readonly string id;
    public Test()
    {
        id = Guid.NewGuid().ToString();
    }

    public string GetId() => id;
}
```

Controller:
```c#
public class WeatherForecastController : ControllerBase
{
     public WeatherForecastController(ILogger<WeatherForecastController> logger,
             ISampleDescription sampleDescription,
             ITest test)
        {
            _logger = logger;
        }
...
}
```

**Comment by alistairjevans** [comment]
Just to add some additional context for this based on the gitter conversation that led to this issue:

- This issue only appears with the multi-tenant container. It does not occur if the regular ``AutofacServiceProviderFactory`` is used instead.
- This issue only appears when using the delegate registration combination:
```csharp
        services.AddScoped<ITest>(sp =>
        {
            return sp.GetRequiredService<Test>();
        });

        services.AddScoped<Test>();
```
If the non-delegate form is used it doesn't happen:
```csharp
services.AddScoped<ITest, Test>();
```

**Comment by cpi01** [comment]
Another behavior that i noticed is when i have types by tenant registered on multitenant container, they cant be resolved at runtime, unless they are registered as singleinstance. Check the following examaple:

**This works**

```
            mtc.ConfigureTenant(Domain.Consts.Tenant.SKYSTOREGB,
               b => b
                   .RegisterType<SampleDescriptionGB>()
                   .As<ISampleDescription>()
                   .SingleInstance());
 mtc.ConfigureTenant(Domain.Consts.Tenant.SKYSTOREDE,
               b => b
                   .RegisterType<SampleDescriptionDE>()
                   .As<ISampleDescription>()
                   .SingleInstance());
            return mtc;    
```      

**This not work**

```
            mtc.ConfigureTenant(Domain.Consts.Tenant.SKYSTOREGB,
               b => b
                   .RegisterType<SampleDescriptionGB>()
                   .As<ISampleDescription>()
                   .InstancePerLifetimeScope());

            mtc.ConfigureTenant(Domain.Consts.Tenant.SKYSTOREDE,
               b => b
                   .RegisterType<SampleDescriptionDE>()
                   .As<ISampleDescription>()
                   .InstancePerLifetimeScope());
            return mtc;
```

**Comment by tillig** [comment]
@alsami you've most recently spent time in here, do you have time to check into this?

@lillo42 @cpi01 while I recognize you're looking for help, any assistance you can provide in actually stepping in with a debugger and locating the root cause of the issue will get your challenges solved faster. We do our best, but there are lots of questions and we get spread thin.

**Comment by alsami** [comment]
@tillig I'll try to look into it tommorow.

**Comment by alsami** [comment]
That's a tough one. It's definitely related to this package and not a combination of newer versions of `Autofac` or what-ever. 

I tried it with a project using the latest version and with a version before `NETCore-App 3` compatibility. The older version works perfectly fine even when updated peer dependencies manually (`Autofac`, `Autofac.Multitenant`, `Autofac.Extensions.DependencyInjection`).

Some what the resolve operation behaves differently with the latest version on the second request.

I will look into it again somewhat later.

**Comment by tillig** [comment]
Resolved by #24, released in [3.0.1](https://github.com/autofac/Autofac.AspNetCore.Multitenant/releases/tag/v3.0.1).
